lib/vagrant_cloud/client.rb

```diff
@@ -1,74 +1,454 @@
-require 'json'
-
 module VagrantCloud
   class Client
+    include Logger
     # Base Vagrant Cloud API URL
-    URL_BASE = 'https://vagrantcloud.com/api/v1'.freeze
-    attr_accessor :access_token
-    attr_accessor :url_base
-
-    # @param [String] access_token - token used to authenticate API requests
-    # @param [String] url_base - URL used to make API requests
-    def initialize(access_token = nil, url_base = nil)
-      if url_base
-        @url_base = url_base
-      else
-        @url_base = URL_BASE
-      end
+    DEFAULT_URL = 'https://vagrantcloud.com/api/v1'.freeze
+    # Valid methods that can be retried
+    IDEMPOTENT_METHODS = [:get, :head].freeze
+    # Number or allowed retries
+    IDEMPOTENT_RETRIES = 3
+    # Number of seconds to wait between retries
+    IDEMPOTENT_RETRY_INTERVAL = 2
+    # Methods which require query parameters
+    QUERY_PARAMS_METHODS = [:get, :head, :delete].freeze
+    # Default instrumentor
+    DEFAULT_INSTRUMENTOR = Instrumentor::Collection.new
 
-      @access_token = access_token
+    # @return [Instrumentor::Collection]
+    def self.instrumentor
+      DEFAULT_INSTRUMENTOR
     end
 
-    # @param [String] method
-    # @param [String] path
-    # @param [Hash] params
-    # @param [String] token
-    # @return [Hash]
-    def request(method, path, params = {}, token = nil)
-      headers = {}
+    # @return [String] Access token for Vagrant Cloud
+    attr_reader :access_token
+    # @return [String] Base request path
+    attr_reader :path_base
+    # @return [String] URL for initializing connection
+    attr_reader :url_base
+    # @return [Integer] Number of retries on idempotent requests
+    attr_reader :retry_count
+    # @return [Integer] Number of seconds to wait between requests
+    attr_reader :retry_interval
+    # @return [Instrumentor::Collection] Instrumentor in use
+    attr_reader :instrumentor
 
-      if token
-        headers['Authorization'] = "Bearer #{token}"
-      elsif @access_token
-        headers['Authorization'] = "Bearer #{@access_token}"
+    # Create a new Client instance
+    #
+    # @param [String] access_token Authentication token for API requests
+    # @param [String] url_base URL used to make API requests
+    # @param [Integer] retry_count Number of retries on idempotent requests
+    # @param [Integer] retry_interval Number of seconds to wait between requests
+    # @param [Instrumentor::Core] instrumentor Instrumentor to use
+    # @return [Client]
+    def initialize(access_token: nil, url_base: nil, retry_count: nil, retry_interval: nil, instrumentor: nil)
```

Comment:
Are we able to leave off `nil` here like some of the other methods and just do something like `initialize(access_token:, url_base:,.....etc...)`?

Reply:
Since the `access_token` is not required, we want to set it to `nil` as default